src/signMessage.c

```diff
@@ -267,22 +274,76 @@ void handle_sign_message_parse_message(volatile unsigned int *tx) {
     }
 }
 
-void handle_sign_message_UI(volatile unsigned int *flags) {
+static bool check_swap_validity(const SummaryItemKind_t kinds[MAX_TRANSACTION_SUMMARY_ITEMS],
+                                size_t num_summary_steps) {
+    bool amount_ok = false;
+    bool recipient_ok = false;
+    bool fee_payer_ok = false;
+    if (num_summary_steps != 3) {
+        PRINTF("3 steps expected for transaction in swap context, not %u\n", num_summary_steps);
+        return false;
+    }
+    for (size_t i = 0; i < num_summary_steps; ++i) {
+        transaction_summary_display_item(i, DisplayFlagNone | DisplayFlagLongPubkeys);
+        PRINTF("G_transaction_summary_title = %s\n", G_transaction_summary_title);
+        PRINTF("G_transaction_summary_text = %s\n", G_transaction_summary_text);
+        switch (kinds[i]) {
+            case SummaryItemAmount:
+                if (strcmp(G_transaction_summary_title, "Transfer") == 0) {
+                    amount_ok = check_swap_amount(G_transaction_summary_text);
+                } else {
+                    PRINTF("Refused field '%s'\n", G_transaction_summary_title);
+                    return false;
+                }
+                break;
+            case SummaryItemPubkey:
+                if (strcmp(G_transaction_summary_title, "Recipient") == 0) {
+                    recipient_ok = check_swap_recipient(G_transaction_summary_text);
+                } else if (strcmp(G_transaction_summary_title, "Fee payer") == 0) {
+                    fee_payer_ok = check_swap_fee_payer(G_transaction_summary_text);
+                } else {
+                    PRINTF("Refused field '%s'\n", G_transaction_summary_title);
+                    return false;
+                }
+                break;
+            default:
+                PRINTF("Refused kind '%u'\n", SummaryItemAmount);
+                return false;
+        }
+    }
+    return amount_ok && recipient_ok && fee_payer_ok;
+}
+
+void handle_sign_message_UI(bool called_from_swap, volatile unsigned int *flags) {
     // Display the transaction summary
     SummaryItemKind_t summary_step_kinds[MAX_TRANSACTION_SUMMARY_ITEMS];
     size_t num_summary_steps = 0;
     if (transaction_summary_finalize(summary_step_kinds, &num_summary_steps) == 0) {
-        size_t num_flow_steps = 0;
+        // If we are in swap context, do not redisplay the message data
+        // Instead, ensure they are identitical with what was previously displayed
+        if (called_from_swap) {
+            if (check_swap_validity(summary_step_kinds, num_summary_steps)) {
+                send_result_sign_message();
+                // Quit app, we are in limited mode and our work is done
+                os_sched_exit(0);
+            } else {
+                PRINTF("Refused blind signing incorrect Swap transaction\n");
+                sendResponse(0, false);
```

Comment:
for some reason calling sendResponse(0, false); disn't work well for me when called not from UX steps, I can try to reproduce. also it doesn't return immediately, so code will continue

Reply:
I've replaced this send() by a throw so we quit the app without going throw `sendResponse()` which sends us to `ui_idle()`